### PR TITLE
Implement legacy OS booting support

### PR DESCRIPTION
**Pull request by Goldfish64**
Adds support for booting of legacy OSes such as older versions of Windows or situations where legacy Windows must be used for proper functionality of hardware. Scope includes both Macs (via legacy interface) and OpenDuet on PC systems.

Closes https://github.com/acidanthera/bugtracker/issues/912.

**Mac**
- [x] Apple legacy loader interface
- [x] Legacy CD-ROM booting
- [x] Get default entry for legacy entries
- [x] Set default entry for legacy entries
- [x] Set desired partition as active to support multiple legacy OSes on a single disk

**PC**
- [x] Legacy loader interface on OpenDuet and firmwares with CSM
- [x] ~~Make USB be in usable state for real mode (transition ownership back to BIOS)~~ Use BlockIo build of DUET
- [x] Get default entry for legacy entries (translate Apple legacy interface path set by macOS)
- [x] Set default entry for legacy entries

**PC and Mac**
- [x] Move legacy interfaces to separate driver
- [x] Configuration to exclude legacy devices from boot
- [x] .contentDetails support if appropriate filesystem drivers are available (NTFS, etc)

### CODE REVIEW

**Comment by joevt** [comment]
> Set desired partition as active to support multiple legacy OSes on a single disk from within OpenCore on Macs

This feature exists in rEFInd. I'm glad to see it coming to OpenCore. It involves writing to the MBR to change the active partition. This assumes the boot code in the MBR supports the MBR partition table semantics but I don't know of a case where an MBR wouldn't have a partition table.

File: Library/DuetBdsLib/BdsPlatform.c [hunk 1]
```diff
@@ -1091,14 +974,6 @@ Routine Description:
     DetectAndPreparePlatformPciDevicePath (TRUE);
   }
 
-  //
-  // The ConIn devices connection will start the USB bus, should disable all
-  // Usb legacy support firstly.
-  // Caution: Must ensure the PCI bus driver has been started. Since the
-  // ConnectRootBridge() will create all the PciIo protocol, it's safe here now
-  //
-  DisableUsbLegacySupport ();
```
Comment (vit9696) [inline Library/DuetBdsLib/BdsPlatform.c:1100]: Is it really not needed?

File: Library/OcBootManagementLib/BootEntryManagement.c [hunk 2]
```diff
@@ -714,12 +714,25 @@ InternalAddBootEntryFromCustomEntry (
     DEBUG_INFO,
     "OCB: Adding custom entry %s (%a|B:%d) -> %a\n",
     BootEntry->Name,
-    CustomEntry->SystemAction != NULL ? "action" : (CustomEntry->Tool ? "tool" : "os"),
+    CustomEntry->ExternalSystemAction != NULL ? "ext-action" : (CustomEntry->SystemAction != NULL ? "action" : (CustomEntry->Tool ? "tool" : "os")),
```
Comment (vit9696) [inline Library/OcBootManagementLib/BootEntryManagement.c:717]: Does this need icon update in OcBinaryData?

File: Library/OcFileLib/DiskMisc.c [hunk 3]
```diff
@@ -636,3 +823,385 @@ OcGetGptPartitionEntry (
...
+  // Protective MBR is defined as a single partition of type 0xEE, other three partitions are to be zero.
+  //
+  if (Mbr->Signature != MBR_SIGNATURE) {
+    FreePool (Mbr);
+    return NULL;
+  }
+
+  if (CheckPartitions) {
+    if (  (Mbr->Partition[0].OSIndicator == PMBR_GPT_PARTITION)
+       && (*((UINT32 *)Mbr->Partition[0].StartingLBA) == 0x01)
```
Comment (vit9696) [inline Library/OcFileLib/DiskMisc.c:884]: Use ReadUnaligned32 just in case.

**Comment by vit9696** [comment]
This looks architecturally right to me. Let's get it merged in the September release, cc @mhaeuser

File: Library/DuetBdsLib/BdsPlatform.c (hunk 1 above)
Comment (Goldfish64) [reply Library/DuetBdsLib/BdsPlatform.c:1100]: Removed in favor of the PCDs checked by UhciDxe and EhciDxe. This prevents legacy USB from being turned off on the non-USB DUET (BlockIo).

File: Library/OcBootManagementLib/BootEntryManagement.c (hunk 2 above)
Comment (Goldfish64) [reply Library/OcBootManagementLib/BootEntryManagement.c:717]: Flavor still needs to be set like for SystemAction, currently OC_FLAVOUR_WINDOWS is used for all entries at the moment. Shouldn't need any icon updates.

**Comment by vit9696** [comment]
In my opinion, this is ready. @Goldfish64, can you merge it please?

**Comment by startergo** [comment]
I have just tested it. The Windows (Legacy) logo appears in the bootlicker, but it can't boot. It boots fine through the Apple BootPicker menu.
[opencore-2023-09-08-191415.txt.zip](https://github.com/acidanthera/OpenCorePkg/files/12563035/opencore-2023-09-08-191415.txt.zip)

**Pull request merged**